zyte_api/aio/client.py

```diff
@@ -146,10 +146,20 @@ def request_parallel_as_completed(self,
         """
         sem = asyncio.Semaphore(self.n_conn)
 
+        if session is not None:
+            close_session = False
+        else:
+            session = create_session(n_conn=self.n_conn)
+            close_session = True
+
         async def _request(query):
             async with sem:
                 return await self.request_raw(query,
                     endpoint=endpoint,
                     session=session)
 
-        return asyncio.as_completed([_request(query) for query in queries])
+        try:
+            yield from asyncio.as_completed([_request(query) for query in queries])
+        finally:
+            if close_session:
+                asyncio.run(session.close())
```

Comment:
How did you check that the session closing works properly?
Im asking because it looks tricky. For example, if this code returns futures, then the session is going to be closed when the last future is returned. But in the original code example the session was closed when the last future was awaited.

Reply:
Tricky indeed, and I did not test it.

In that case, I see 2 options:
- We do not implement a default session, and instead we implement something like #22 (I was working on that already). In the `iter` example we can feature no session to keep the code simple, but in the docstrings of both `get` and `iter` we recommend using the new session context manager.
- We go back to making this an async iterator, await before yielding, yield exceptions as in the sync API, and then we can close knowing that everything has been awaited.

Reply:
Went with 2 here.